### PR TITLE
:sparkles: upgrade to k8s 0.27.1

### PR DESCRIPTION
**Pull request by mythi**
This PR moves k8s dependencies to 0.27.1 and reworks things to make the tests/linter happy.

k8s.io/apimachinery deprecated legacy Poll methods and this
triggers golangi-lint staticheck errors.
    
Rework all the poll logic to move to the preferred methods. 

Details: https://github.com/kubernetes/apimachinery/commit/831cf05a7c4f839939a6c5e1e8bee238ddc6baac

Additionally, NamespacedName type now implements MarshalLog() which
replaces the functionality in kube_helpers.go. See k/k [#117238.](https://github.com/kubernetes/kubernetes/pull/117238).

### CODE REVIEW

**Comment by howardjohn** [comment]
I have an alternative PR that makes controller-runtime support 1.27 *and* 1.26 concurrently so projects do not need to serialize updates: https://github.com/kubernetes-sigs/controller-runtime/pull/2223

I am also fine with this, just presenting an option. I would like to see some support for 1.27 merged ideally, though.

**Comment by mythi** [comment]
> I have an alternative PR that makes controller-runtime support 1.27 _and_ 1.26 concurrently so projects do not need to serialize updates: #2223
> 
> I am also fine with this, just presenting an option. I would like to see some support for 1.27 merged ideally, though.

I'm totally fine with the proposed alternative. My goal (this PR) was also to just get to 1.27...

**Comment by mythi** [comment]
> > I have an alternative PR that makes controller-runtime support 1.27 _and_ 1.26 concurrently so projects do not need to serialize updates: #2223
> > I am also fine with this, just presenting an option. I would like to see some support for 1.27 merged ideally, though.
> 
> I'm totally fine with the proposed alternative. My goal (this PR) was also to just get to 1.27...

Rebased since #2223 is now merged.

**Comment by mythi** [comment]
@alvaroaleman @vincepri could I get `ok-to-test` to see tests are passing. `make test` was passing for me locally

File: pkg/certwatcher/certwatcher.go
Comment (alvaroaleman) [inline pkg/certwatcher/certwatcher.go:95]: I think we should always return an err here if we encountered one, not only when its DeadlineExceeded

File: pkg/envtest/crd.go
Comment (alvaroaleman) [inline pkg/envtest/crd.go:169]: Why was this changed?

File: pkg/envtest/webhook.go
Comment (alvaroaleman) [inline pkg/envtest/webhook.go:150]: Why was this added?

File: pkg/envtest/webhook.go
Comment (alvaroaleman) [inline pkg/envtest/webhook.go:218]: Why was this changed?

File: pkg/internal/source/kind.go
Comment (alvaroaleman) [inline pkg/internal/source/kind.go:56]: Why was this changed?

File: pkg/log/zap/zap_test.go
Comment (alvaroaleman) [inline pkg/log/zap/zap_test.go:221]: Why is this needed?

File: pkg/log/zap/zap_test.go
Comment (alvaroaleman) [inline pkg/log/zap/zap_test.go:256]: Why?

File: pkg/manager/internal.go
Comment (alvaroaleman) [inline pkg/manager/internal.go:432]: Why is this needed?

**Review by alvaroaleman — COMMENTED** [review]
There is a bunch of unrelated changed in here, please keep this purely to updating the deps and do a second PR for any other change

File: pkg/log/zap/zap_test.go
Comment (mythi) [reply pkg/log/zap/zap_test.go]: Sorry, this can be ignored and I will fix it properly. Currently something in 0.27 breaks marshaling and these tests fail (see the CI results)

**Comment by mythi** [comment]
> There is a bunch of unrelated changed in here, please keep this purely to updating the deps and do a second PR for any other change

I updated the PR cover letter from the commit message to explain why many of these changes are needed. Updating is not possible without the linter fixes.

File: pkg/certwatcher/certwatcher.go
Comment (mythi) [reply pkg/certwatcher/certwatcher.go:95]: will change to `wait.Interrupted(err)`. the context func does not seem to any return errors.

**Comment by mythi** [comment]
> > There is a bunch of unrelated changed in here, please keep this purely to updating the deps and do a second PR for any other change
> 
> I updated the PR cover letter from the commit message to explain why many of these changes are needed. Updating is not possible without the linter fixes.

@alvaroaleman did I cover all your questions? However, this PR is blocked until https://github.com/kubernetes/kubernetes/pull/117238 is released.

File: pkg/internal/source/kind.go
Comment (AliDatadog) [reply pkg/internal/source/kind.go:56]: `PollImmediateUntilWithContext` is deprecated

**Comment by sbueringer** [comment]
Another release? 0.27.1 didn't have the fix, or we need another one?

**Comment by mythi** [comment]
> Another release? 0.27.1 didn't have the fix, or we need another one?

The backport/cherry-pick has not been merged yet: https://github.com/kubernetes/kubernetes/pull/117298

**Comment by mythi** [comment]
> > Another release? 0.27.1 didn't have the fix, or we need another one?
> 
> The backport/cherry-pick has not been merged yet: [kubernetes/kubernetes#117298](https://github.com/kubernetes/kubernetes/pull/117298)

but people looking to update their dependencies to k8s 0.27 seem to be hitting the build error that #2223 fixes so maybe that alone would be sufficient until we can get this merged? that allowed me to (test) build my controller/webhook.

File: pkg/internal/source/kind.go
Comment (mythi) [reply pkg/internal/source/kind.go:56]: AFAICS `PollWithContextCancel` is not defined. I wonder if it's a typo because the commit adds `PollUntilContextCancel` https://github.com/kubernetes/apimachinery/commit/831cf05a7c4f839939a6c5e1e8bee238ddc6baac

Comment (camilamacedo86) [reply pkg/internal/source/kind.go:56]: It seems a typo issue.

File: pkg/certwatcher/certwatcher.go
Comment (camilamacedo86) [inline pkg/certwatcher/certwatcher.go:94]: why this change is required?

File: pkg/log/zap/kube_helpers.go
Comment (camilamacedo86) [inline pkg/log/zap/kube_helpers.go:57]: Why are you removing the helpers? 
It does not seem  required  at all.

File: pkg/log/zap/kube_helpers.go
Comment (camilamacedo86) [inline pkg/log/zap/kube_helpers.go:122]: same here. why remove the helper?

File: pkg/manager/internal.go
Comment (camilamacedo86) [inline pkg/manager/internal.go:468]: The deprecated message suggest use here wait. Interrupted instead. So, I think we should use it. 
We also could here check if the context was canceled or exceeded.

File: go.mod
Comment (camilamacedo86) [inline go.mod:17]: Why bump k8s is a breaking change for CR and not ✨ ?

**Comment by vincepri** [comment]
/hold

Going to hold until we can audit the polling deprecations and other changes within this PR, once maintainers have some time after KubeCon.

File: pkg/log/zap/kube_helpers.go
Comment (mythi) [reply pkg/log/zap/kube_helpers.go:57]: This part of the code never gets called after the k8s patch and the end result becomes the same (MarshalLog() returns lowercase name/namespacename)

File: pkg/manager/internal.go
Comment (mythi) [reply pkg/manager/internal.go:468]: This was following the item 2 in the migration note but I can change to wait.Interrupted

File: go.mod
Comment (mythi) [reply go.mod:17]: It used to be but #2223 addressed the breakage properly so I can remove ✨

Comment (camilamacedo86) [reply go.mod:17]: Sorry, are the end-user affected by those changes? Will they need to change their code when they bump a new release within this? If not, then IMO it seems that the right emoji is  `:sparkles:`

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
It is falling in the tests: 
Screenshot 2023-04-19 at 18 03 23

The root cause of the above error is that the log name and namespace are now outputted with UpperCase (`Name` and `Namespace`), see: https://github.com/kubernetes-sigs/controller-runtime/pull/2281/files#diff-a985354f6767137e1c7c14ae824c6048d1f58b926c041b41e0f68465729ec7eb

File: go.mod
Comment (mythi) [reply go.mod:17]: changed

**Comment by mythi** [comment]
> The root cause of the above error is that the log name and namespace are now outputted with UpperCase

The test case catched a real issue and that was fixed in k8s. I root-caused the issue too late and 0.27(.1) happened before the fix was available. Changing the test case to get CI green is not the right fix.

File: pkg/manager/internal.go
Comment (mythi) [reply pkg/manager/internal.go:468]: but how I understand the code `.Start()` wait until the context is cancelled or there's an error (i.e., no timeouts?) so `context.Canceled` is a better match

**Comment by camilamacedo86** [comment]
Hi @mythi,

> The test case catched a real issue and that was fixed in k8s. I root-caused the issue too late and 0.27(.1) happened before the fix was available. Changing the test case to get CI green is not the right fix.

Could you please clarify why do you think that it is an issue in the new deps versions and not a change or bug fix?
Have we any raised issues or discussion topics within the project that made the change?

File: pkg/manager/internal.go
Comment (camilamacedo86) [reply pkg/manager/internal.go:468]: Should we not either check for `context.DeadlineExceeded`? 

```go
if !errors.Is(err, context.Canceled) && !errors.Is(err, context.DeadlineExceeded)   {

```

File: pkg/envtest/webhook.go
Comment (joelanford) [reply pkg/envtest/webhook.go:150]: Is this something where the behavior changed because of the k8s bump? Or is this an unrelated bug fix? If the latter, should we split it out into a separate PR?

Comment (mythi) [reply pkg/envtest/webhook.go:150]: Good question. `options.MaxTime` is handled differently in `poll.Immediate()`:

```
// closed. A timeout of 0 is interpreted as an infinity, and in such a case
// it would be the caller's responsibility to close the done channel.
// Failure to do so would result in a leaked goroutine.
```

without `defaultWebhookOptions()`, `MaxTime` is zero. If "inifinity" is not wanted, this change could be moved to a separate PR.

**Comment by zetaab** [comment]
@mythi ideas when this PR will get some progress? There are quite much other repositories depending of this update.

**Comment by mythi** [comment]
> @mythi ideas when this PR will get some progress? There are quite much other repositories depending on this update.

I've noticed that too but to make progress with this PR, we need k8s 1.27.2 with the fixes we depend on. I'm actively following this and able to rebase as soon as it's possible.

**Comment by sbueringer** [comment]
I think it would be good if we can resolve all findings except the one were we depend on 1.27.2 in the meantime

**Comment by mythi** [comment]
> I think it would be good if we can resolve all findings except the one were we depend on 1.27.2 in the meantime

Currently I'm aware of [one comment](https://github.com/kubernetes-sigs/controller-runtime/pull/2189#discussion_r1172080703) I could spin off to another PR but other than that I'm waiting for the maintainer feedback (this was put on hold due to KubeCon last week).

**Comment by sbueringer** [comment]
> > I think it would be good if we can resolve all findings except the one were we depend on 1.27.2 in the meantime
> 
> Currently I'm aware of [one comment](https://github.com/kubernetes-sigs/controller-runtime/pull/2189#discussion_r1172080703) I could spin off to another PR but other than that I'm waiting for the maintainer feedback (this was put on hold due to KubeCon last week).

Sorry my bad, just saw all the open conversations. I'll try to take a closer look today or tomorrow

**Comment by mythi** [comment]
> > > I think it would be good if we can resolve all findings except the one were we depend on 1.27.2 in the meantime
> > 
> > 
> > Currently I'm aware of [one comment](https://github.com/kubernetes-sigs/controller-runtime/pull/2189#discussion_r1172080703) I could spin off to another PR but other than that I'm waiting for the maintainer feedback (this was put on hold due to KubeCon last week).
> 
> Sorry my bad, just saw all the open conversations. I'll try to take a closer look today or tomorrow.
> 
> Can you please rebase to get rid of the merge conflict?

Yeah sorry about the open conversations. I closed some of them now and rebased+force pushed.

File: pkg/envtest/webhook.go
Comment (mythi) [reply pkg/envtest/webhook.go]: Created #2289, closing.

File: pkg/envtest/webhook.go
Comment (mythi) [reply pkg/envtest/webhook.go:206]: answered in the PR description

File: pkg/manager/internal.go
Comment (mythi) [reply pkg/manager/internal.go:461]: answered in the PR description

File: pkg/manager/internal.go
Comment (mythi) [reply pkg/manager/internal.go:468]: `cm.internalCtx` has no deadline AFAICS.

File: pkg/certwatcher/certwatcher.go
Comment (mythi) [reply pkg/certwatcher/certwatcher.go:94]: `err != nil` seems to be for `ErrWaitTimeout` only and the migration note suggests:

```
`Interrupted(err) bool` has been added that should be used instead of
checking `err == ErrWaitTimeout`. 
```

File: pkg/internal/source/kind.go
Comment (sbueringer) [reply pkg/internal/source/kind.go:56]: @mythi Looks like it was incorrect, but they didn't mean `PollUntilContextCancel`, they meant `PollUntilContextTimeout` (https://github.com/kubernetes/kubernetes/commit/1cce3b3e2f5981a6cdc2d8ead921538533399575)

File: pkg/certwatcher/certwatcher.go
Comment (sbueringer) [reply pkg/certwatcher/certwatcher.go:94]: > err != nil seems to be for ErrWaitTimeout

`err != nil` checks for all errors not only ErrWaitTimeout

> `Interrupted(err) bool` has been added that should be used instead of
checking `err == ErrWaitTimeout`. 

We didn't use `err == ErrWaitTimeout` before so not sure why we should use Interrupted now?

File: pkg/envtest/crd.go
Comment (sbueringer) [inline pkg/envtest/crd.go:169]:
```suggestion
	return wait.PollUntilContextTimeout(context.TODO(), options.PollInterval, options.MaxTime, true, p.poll)
```
I think we should use TODO here signaling that sooner or later we should pass through a ctx

File: pkg/internal/source/kind.go
Comment (mythi) [reply pkg/internal/source/kind.go:56]: @sbueringer thanks for the update!

File: pkg/envtest/webhook.go
Comment (sbueringer) [inline pkg/envtest/webhook.go:218]:
```suggestion
	return wait.PollUntilContextTimeout(context.TODO(), options.PollInterval, options.MaxTime, true, p.poll)
```

File: pkg/manager/internal.go
Comment (sbueringer) [reply pkg/manager/internal.go:468]: Consumers of controller-runtime can add their own Runnables to a manager (https://github.com/kubernetes-sigs/controller-runtime/blob/7f57edcf611232b438f2c5700d225c338235f0a4/pkg/manager/runnable_group.go#L53).

We have no control over which wait functions they might use or which errors they return.

I would use the following:
```go
		if !(errors.Is(err, wait.ErrWaitTimeout) || wait.Interrupted(err)) {
			return err
		}
```

`wait.Interrupted` internally checks for the new private version of errWaitTimeout, Canceled and DeadlineExceeded:
```go
func Interrupted(err error) bool {
	switch {
	case errors.Is(err, errWaitTimeout),
		errors.Is(err, context.Canceled),
		errors.Is(err, context.DeadlineExceeded):
		return true
	default:
		return false
	}
}
```

Comment (sbueringer) [reply pkg/manager/internal.go:468]: @vincepri @alvaroaleman Do you remember why we were continuing with the Start func if we got a wait.ErrWaitTimeout?

File: pkg/certwatcher/certwatcher.go
Comment (mythi) [reply pkg/certwatcher/certwatcher.go:94]: I think it's fine to keep `err != nil` but what I wanted to say was the errors we are getting from the new func are different and the errors are covered by `wait.Interrupted()` since the condition never returns errors: "The returned error is the error returned by the last condition or the context error if the context was terminated."

File: pkg/manager/internal.go
Comment (mythi) [reply pkg/manager/internal.go:468]: > `if !(errors.Is(err, wait.ErrWaitTimeout)`

this is also deprecated so it'll trigger a warning

File: pkg/certwatcher/certwatcher.go
Comment (sbueringer) [reply pkg/certwatcher/certwatcher.go]: Sounds good!

File: pkg/manager/internal.go
Comment (sbueringer) [reply pkg/manager/internal.go:468]: That's not a showstopper we can add a no lint (~ `//nolint:<linter-name>`)

**Comment by sbueringer** [comment]
lgtm from my side pending 0.27.2 and resolution of https://github.com/kubernetes-sigs/controller-runtime/pull/2189#discussion_r1171151477 (let's wait for feedback there)

File: pkg/manager/internal.go
Comment (sbueringer) [reply pkg/manager/internal.go:468]: @mythi Talked to Vince/Alvaro. We would change it to simply return all errors 

Also we would like to merge this PR ASAP and just for now skip the log tests.

Delta looks roughly like this: https://github.com/mythi/controller-runtime/compare/27-upgrade...sbueringer:controller-runtime:pr-27-upgrade-adjusted (I used my branch to verify that everything works as expected with Cluster API)

We will still hold the release until we bumped to 1.27.2 though

Comment (sbueringer) [reply pkg/manager/internal.go:468]: Sounds good. Done

Comment (mythi) [reply pkg/manager/internal.go]: Rebased and included all the proposed updates. We should be getting all tests passing :crossed_fingers:

Comment (sbueringer) [reply pkg/manager/internal.go]: Thx!

**Comment by sbueringer** [comment]
@mythi Thank you very much for all your work on this!!

/lgtm

**Review by alvaroaleman — APPROVED** [review]
Thanks!

**Comment by mythi** [comment]
> @mythi Thank you very much for all your work on this!!

no problem. Even if this PR turned out to be WAY bigger task than I originally planned, I don't mind since I believe it gives a good learning for the future as well: it's probably good to start pulling/testing k8s release candidates to ensure all is good ahead of the release.

**Pull request merged**
